### DISCUSSION TITLE
zh/tutorials/Streamlit/streamlit-series-basic-output

### DISCUSSION THREAD

**Discussion by giscus** [TeddyHuang-00/teddyhuang-00.github.io#152] [Announcements]
# zh/tutorials/Streamlit/streamlit-series-basic-output

Streamlit 的文字类以及基础输出

https://teddyhuang-00.github.io/zh/tutorials/Streamlit/streamlit-series-basic-output.html

**Comment by echoflying** [comment]
有没有留意到在处理文本内容时，write和markdown是一样的，问题是单个的换行\n是不会换行的，两个\n\n可以换行。text显示\n可以换行。但text的问题是不能自动换行，长文本只显示一行。

我需要的\n能够换行，而且长文本在文本框内能自动换行显示。貌似没有哪个合用。大神指点一二。

**Reply by TeddyHuang-00** [reply]
这个应该说在markdown语法里面本来一个换行符就不起作用，因此有几种解决方案：
1. 在渲染文本时将
    ```python
    st.write(content)
    ```
    改为
    ```python
    st.write(content.replace("\n", "\n\n"))
    ```
    或者使用更精确的正则表达式捕捉单个换行符并替换，这样的话可以正确使用markdown的格式进行分段换行
2. 根据 `"\n"` 切分文本并过滤掉空元素，逐个使用`st.write` 显示，效果同 1.
3. （如果并不想触发分段）可以将单个的换行符替换为 `"<br/>"`，然后使用`st.write`，并且令 `unsafe_allow_html=True`。这样既可以保留原来的双换行符分段，也可以令单个换行符强制渲染为换行。示例如下：
    ```python
    import re

    # Exactly 1 line break, following and followed by other characters
    single_lb_pattern = re.compile(r"([^\n])\n([^\n])")
    # Replace the line break with a break tag
    single_lb_replacement = r"\1<br/>\2"

    content = "abc\ndef\n\n1234567890zxcvbnmasdfghjklqwertyuiop"
    while single_lb_pattern.search(content):
        content = single_lb_pattern.sub(single_lb_replacement, content)

    st.write(content, unsafe_allow_html=True)
    ```

希望对你有帮助
